// @layerzerolabs/io-devtools/dist/swag/index.js
'use strict';

var React = require('react');
var ink = require('ink');
var InkTable = require('ink-table');
var Gradient = require('ink-gradient');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);
var InkTable__default = /*#__PURE__*/_interopDefault(InkTable);
var Gradient__default = /*#__PURE__*/_interopDefault(Gradient);

// src/swag/printer.tsx
var RecordList = ({
  data = [],
  columns
}) => {
  return /* @__PURE__ */ React__default.default.createElement(React__default.default.Fragment, null, data.map((row, index) => /* @__PURE__ */ React__default.default.createElement(Record, { data: row, columns, key: index })));
};
var Record = ({
  data = {},
  columns = ((_a) => (_a = process.stdout.columns) != null ? _a : 80)()
}) => {
  const padding = 1;
  const labels = Object.keys(data);
  const labelLengths = labels.map(({ length }) => length);
  const maxLabelLength = Math.max(1, ...labelLengths);
  const labelColumnWidth = maxLabelLength + 2 * padding;
  const valueColumnWidth = Math.max(10, columns - labelColumnWidth);
  return /* @__PURE__ */ React__default.default.createElement(
    ink.Box,
    {
      flexDirection: "column",
      borderColor: "grey",
      borderStyle: "single",
      paddingX: 1
    },
    Object.entries(data).map(([key, value]) => {
      return /* @__PURE__ */ React__default.default.createElement(ink.Box, { key }, /* @__PURE__ */ React__default.default.createElement(ink.Box, { width: labelColumnWidth }, /* @__PURE__ */ React__default.default.createElement(ink.Text, { bold: true, wrap: "truncate", color: "magenta" }, key)), /* @__PURE__ */ React__default.default.createElement(ink.Box, { width: valueColumnWidth }, value == null ? /* @__PURE__ */ React__default.default.createElement(ink.Text, { wrap: "wrap", color: "gray" }, "-") : /* @__PURE__ */ React__default.default.createElement(ink.Text, { wrap: "wrap" }, String(value))));
    })
  );
};
var Table = ({ data }) => /* @__PURE__ */ React__default.default.createElement(
  InkTable__default.default,
  {
    data,
    header: ({ children }) => /* @__PURE__ */ React__default.default.createElement(ink.Text, { bold: true, color: "magenta" }, children)
  }
);

// src/swag/printer.tsx
var printLogo = () => {
  const logo = `
    \u256D\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u256E
    \u2502       \u2593\u2593\u2593 LayerZero DevTools \u2593\u2593\u2593        \u2502
    \u2502  \u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550    \u2502
    \u2502          /*\\                            \u2502
    \u2502         /* *\\     BUILD ANYTHING        \u2502
    \u2502         ('v')                           \u2502
    \u2502        //-=-\\\\    \u25B6 OMNICHAIN           \u2502
    \u2502        (\\_=_/)                          \u2502
    \u2502         ^^ ^^                           \u2502
    \u2502  \u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593\u2593  \u2502
    \u2570\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u256F
`;
  console.log(logo);
};
var printRecord = (data) => ink.render(/* @__PURE__ */ React__default.default.createElement(Record, { data })).unmount();
var printRecords = (data) => ink.render(/* @__PURE__ */ React__default.default.createElement(RecordList, { data })).unmount();
var printVerticalTable = (data) => ink.render(/* @__PURE__ */ React__default.default.createElement(Table, { data })).unmount();
var ProgressBar = ({
  columns = ((_a) => (_a = process.stdout.columns) != null ? _a : 80)(),
  progress = 0,
  before = "",
  after = ""
}) => {
  const padding = before.length + after.length;
  const space = Math.max(0, columns - padding);
  const clampedProgress = Math.max(0, Math.min(1, progress));
  const numChars = Math.min(Math.floor(space * clampedProgress), space);
  const chars = CHARACTER.repeat(numChars);
  const numEmptyChars = space - numChars;
  const emptyChars = EMPTY_CHARACTER.repeat(numEmptyChars);
  return /* @__PURE__ */ React__default.default.createElement(ink.Box, null, /* @__PURE__ */ React__default.default.createElement(ink.Text, null, before), /* @__PURE__ */ React__default.default.createElement(ink.Box, { alignItems: "center" }, /* @__PURE__ */ React__default.default.createElement(Gradient__default.default, { name: "rainbow" }, /* @__PURE__ */ React__default.default.createElement(ink.Text, null, chars)), /* @__PURE__ */ React__default.default.createElement(ink.Text, null, emptyChars)), /* @__PURE__ */ React__default.default.createElement(ink.Text, null, after));
};
var CHARACTER = "\u2588";
var EMPTY_CHARACTER = "\u2591";
var PENGUIN_FRAMES = [
  [
    `     /*\\       `,
    `    /* *\\     `,
    `    ('-')    `,
    `   //-=-\\\\   `,
    `   (\\_=_/)   `,
    `    ^^ ^^    `
  ],
  [
    ` *   /*\\   *   `,
    `    /* *\\     `,
    `    ('v')    `,
    `   //-=-\\\\-  `,
    `   (\\_=_/)   `,
    `  * ^^ ^^ *  `
  ],
  [
    `   + /*\\ +     `,
    `    /* *\\     `,
    `    ('>')    `,
    `   //-=-\\\\/  `,
    `   (\\_=_/)   `,
    `  + ^^ ^^ +  `
  ],
  [
    `  *  /*\\       `,
    `    /* *\\     `,
    `    ('^')  *`,
    `   //-=-\\\\/  `,
    `   (\\_=_/)   `,
    ` *  ^^ ^^  +  `
  ],
  [
    `  +  /*\\       `,
    `    /* *\\   *`,
    `    ('<')  *`,
    `   //-=-\\\\/  `,
    `   (\\_=_/)   `,
    ` +  ^^ ^^  *  `
  ],
  [
    `  *  /*\\     *`,
    `    /* *\\   *`,
    `    ('v')  *`,
    `   //-=-\\\\/  `,
    `   (\\_=_/)   `,
    ` *  ^^ ^^  +  `
  ],
  [
    ` o*  /*\\     *`,
    `    /* *\\   *`,
    `    ('o')  *`,
    `   //-=-\\\\-  `,
    `   (\\_=_/)   `,
    `  + ^^ ^^ +  `
  ]
];
var PenguinSpinner = ({
  before = "",
  after = "",
  title = "Sending Transactions"
}) => {
  const [frameIndex, setFrameIndex] = React.useState(0);
  React.useEffect(() => {
    const id = setInterval(() => {
      setFrameIndex((prev) => (prev + 1) % PENGUIN_FRAMES.length);
    }, 500);
    return () => clearInterval(id);
  }, []);
  const currentFrame = PENGUIN_FRAMES[frameIndex] || PENGUIN_FRAMES[0];
  return /* @__PURE__ */ React__default.default.createElement(ink.Box, { flexDirection: "column" }, /* @__PURE__ */ React__default.default.createElement(ink.Text, { bold: true }, before, title, after), /* @__PURE__ */ React__default.default.createElement(ink.Box, { flexDirection: "column", marginTop: 1, marginBottom: 1 }, currentFrame.map((line, index) => /* @__PURE__ */ React__default.default.createElement(ink.Text, { key: index }, line))));
};
var createProgressBar = (props = {}) => /* @__PURE__ */ React__default.default.createElement(ProgressBar, { ...props });
var createPenguinSpinner = (props = {}) => /* @__PURE__ */ React__default.default.createElement(PenguinSpinner, { ...props });

Object.defineProperty(exports, "render", {
  enumerable: true,
  get: function () { return ink.render; }
});
exports.createPenguinSpinner = createPenguinSpinner;
exports.createProgressBar = createProgressBar;
exports.printLogo = printLogo;
exports.printRecord = printRecord;
exports.printRecords = printRecords;
exports.printVerticalTable = printVerticalTable;
